### PR TITLE
fix(generator): more cleanup to plugin framework

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
Changes:
- Add "dev" mode to npm scripts for automatic rebuilding to replace linking to source files. This removes the need for having a Webpack alias.
- Clean up `README.md` and add instructions for adding the plugin to Superset instead of as an embedded `SuperChart`
- Add new docs section for creating new viz plugins

### TESTING INSTRUCTIONS
1. go into the generator directory: `cd `superset-frontend/packages/generator-superset`
2. run `npm link`
3. create a new plugin directory: `cd ~/src && mkdir superset-plugin-hello-world`
4. initialize the plugin: `yo @superset-ui/superset`
5. Follow the instructions in the `README.md` file and see if you can make your plugin work

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/packages/generator-superset/generators/plugin-chart/templates/README.erb [hunk 1]
```diff
@@ -1,52 +1,40 @@
...
+To run the plugin in development mode (=rebuilding whenever changes are made), start the dev server with the following command:
 
-new <%= packageLabel %>ChartPlugin()
-  .configure({ key: '<%= packageName %>' })
-  .register();
+```
+npm run dev
+```
+
+To add the package to Superset, go to the `superset-frontend` subdirectory in your Superset source folder (assuming both the `superset-plugin-chart-liquid` plugin and `superset` repos are in the same root directory) and run
```
Comment (geido) [inline superset-frontend/packages/generator-superset/generators/plugin-chart/templates/README.erb:20]: Hey @villebro I am not sure what `superset-plugin-chart-liquid` is referring to. Is this just a sample name?

Comment (villebro) [reply superset-frontend/packages/generator-superset/generators/plugin-chart/templates/README.erb:20]: thanks for catching! It's an example plugin that I've worked on while testing this template

File: superset-frontend/packages/generator-superset/generators/plugin-chart/templates/README.erb [hunk 2]
```diff
@@ -1,52 +1,40 @@
...
 
 ### Usage
 
-Configure `key`, which can be any `string`, and register the plugin. This `key` will be used to lookup this chart throughout the app.
+To build the plugin, run the following commands:
 
-```js
-import <%= packageLabel %>ChartPlugin from '@superset-ui/plugin-chart-<%= packageName %>';
+```
+npm i --force
```
Comment (villebro) [inline superset-frontend/packages/generator-superset/generators/plugin-chart/templates/README.erb:10]: this is unfortunately needed due to the conflicting versions we currently have in our deps (React, AntD etc) 😏

File: superset-frontend/webpack.config.js [hunk 3]
```diff
@@ -285,7 +285,9 @@ const config = {
     // resolve modules from `/superset_frontend/node_modules` and `/superset_frontend`
     modules: ['node_modules', APP_DIR],
     alias: {
-      // TODO: remove alias once React has been upgraaded to v. 17
+      // TODO: remove aliases once React has been upgraaded to v. 17 and
```
Comment (kgabryje) [inline superset-frontend/webpack.config.js:288]:
```suggestion
      // TODO: remove aliases once React has been upgraded to v. 17 and
```

File: superset-frontend/packages/generator-superset/generators/plugin-chart/templates/package.erb [hunk 4]
```diff
@@ -1,46 +1,43 @@
...
   },
-  "dependencies": {
-  },
+  "dependencies": {},
   "peerDependencies": {
-    "react": "^16.13.1",
-    "@superset-ui/core": "*",
-    "@superset-ui/chart-controls": "*"
+    "@superset-ui/chart-controls": "^0.18.25",
+    "@superset-ui/core": "^0.18.25",
```
Comment (zhaoyongjie) [inline superset-frontend/packages/generator-superset/generators/plugin-chart/templates/package.erb:32]: Can we use "*" to adapt the  core plugin?

**Review by zhaoyongjie — APPROVED** [review]
image

I have tested in local environment. LGTM!

**Comment by zhaoyongjie** [comment]
Add some instructions on how to use this awesome plugin system:
1. Install `yo` in your local machine
```shell
npm i -g yo
```
2. Install and link `@superset-ui/generator-superset`
```shell
cd superset-frontend/packages/generator-superset
npm i
npm link
```

3. Check if the installation is successful
```
$ npm list -g
/Users/user-name/.nvm/versions/node/v16.9.1/lib
├── @superset-ui/generator-superset@0.18.25 -> ./../../../../../workspace/warehouse/superset/incubator-superset/superset-frontend/packages/generator-superset
├── corepack@0.9.0
├── npm@7.21.1
├── po2json@0.4.5
├── yarn@1.22.17
└── yo@4.3.0
```

4. Create a plugin directory
```
mkdir -p /tmp/superset-plugin-hello-world
cd /tmp/superset-plugin-hello-world
```

5. initialize the plugin
```
yo @superset-ui/superset
```

6. install and build `superset-plugin-hello-world`
```
npm i --force
npm run build
```

7. Install `superset-plugin-hello-world` into Superset
```
npm i -S /tmp/superset-plugin-hello-world
```

8. load customize the plugin. This is git-diff in my local machine
```
(superset) yongjie.zhao@:incubator-superset$ git diff
diff --git a/superset-frontend/package.json b/superset-frontend/package.json
index 439322dce..d9b87130b 100644
--- a/superset-frontend/package.json
+++ b/superset-frontend/package.json
@@ -195,6 +195,7 @@
     "rison": "^0.1.1",
     "scroll-into-view-if-needed": "^2.2.28",
     "shortid": "^2.2.6",
+    "superset-plugin-hello-world": "file:../../../../../superset-plugin-hello-world",
     "urijs": "^1.19.6",
     "use-immer": "^0.6.0",
     "use-query-params": "^1.1.9",
diff --git a/superset-frontend/src/visualizations/presets/MainPreset.js b/superset-frontend/src/visualizations/presets/MainPreset.js
index a32fbb2b7..1318bbb9c 100644
--- a/superset-frontend/src/visualizations/presets/MainPreset.js
+++ b/superset-frontend/src/visualizations/presets/MainPreset.js
@@ -80,6 +80,7 @@ import {
 import { PivotTableChartPlugin as PivotTableChartPluginV2 } from '@superset-ui/plugin-chart-pivot-table';
 import FilterBoxChartPlugin from '../FilterBox/FilterBoxChartPlugin';
 import TimeTableChartPlugin from '../TimeTable/TimeTableChartPlugin';
+import { SupersetPluginHelloWorld } from 'superset-plugin-hello-world';

 export default class MainPreset extends Preset {
   constructor() {
@@ -164,6 +165,7 @@ export default class MainPreset extends Preset {
         new TimeColumnFilterPlugin().configure({ key: 'filter_timecolumn' }),
         new TimeGrainFilterPlugin().configure({ key: 'filter_timegrain' }),
         new EchartsTreeChartPlugin().configure({ key: 'tree_chart' }),
+        new SupersetPluginHelloWorld().configure({ key: 'hello_world'}),
         ...experimentalplugins,
       ],
     });
```

9. run development server in Superset
```
npm run dev-server
```

File: superset-frontend/packages/generator-superset/generators/plugin-chart/templates/README.erb (hunk 2 above)
Comment (kgabryje) [reply superset-frontend/packages/generator-superset/generators/plugin-chart/templates/README.erb:10]: We can also use `--legacy-peer-deps`, looks a bit less aggressive 😛

**Comment by rusackas** [comment]
Looks like the main readme [here](https://github.com/apache/superset/tree/master/superset-frontend/packages/generator-superset) needs some love, too. It still references `superset-ui` and all of that :D

File: superset-frontend/packages/generator-superset/generators/plugin-chart/templates/README.erb (hunk 2 above)
Comment (villebro) [reply superset-frontend/packages/generator-superset/generators/plugin-chart/templates/README.erb:10]: With `--legacy-peer-deps`, the React and the core packages won't be present due to them being peer dependencies, so running tests will fail:

```bash
npm run test
```

```
src/types.ts:19:68 - error TS2307: Cannot find module '@superset-ui/core' or its corresponding type declarations.

19 import { QueryFormData, supersetTheme, TimeseriesDataRecord } from '@superset-ui/core';
                                                                      ~~~~~~~~~~~~~~~~~~~

src/SupersetPluginChartHelloWorld.tsx:19:45 - error TS2307: Cannot find module 'react' or its corresponding type declarations.

19 import React, { useEffect, createRef } from 'react';
                                               ~~~~~~~
```

**Pull request merged**